waspc/waspls/src/Wasp/LSP/TypeInference.hs

```diff
@@ -0,0 +1,132 @@
+module Wasp.LSP.TypeInference
+  ( -- * Inferred types for CST locations
+    inferTypeAtLocation,
+
+    -- * Lower level pieces
+    ExprPath,
+    ExprKey (..),
+    findExprPathAtLocation,
+    findTypeForPath,
+  )
+where
+
+import Control.Monad (guard)
+import Data.Foldable (find)
+import qualified Data.HashMap.Strict as M
+import qualified Wasp.Analyzer.Parser.CST as S
+import Wasp.Analyzer.Parser.CST.Traverse (Traversal)
+import qualified Wasp.Analyzer.Parser.CST.Traverse as T
+import Wasp.Analyzer.StdTypeDefinitions (stdTypes)
+import Wasp.Analyzer.Type (Type)
+import qualified Wasp.Analyzer.Type as Type
+import Wasp.Analyzer.TypeDefinitions (DeclType (dtBodyType), getDeclType)
+import Wasp.LSP.Syntax (lexemeAt)
+
+-- | Infer the type at a location in the CST, or 'Nothing' if type could not be
+-- determined.
+inferTypeAtLocation :: String -> Traversal -> Maybe Type
+inferTypeAtLocation src location = case findExprPathAtLocation src location of
+  Just path -> findTypeForPath path
+  _ -> Nothing
+
+-- | A "path" through wasp expressions to a certain location.
+--
+-- === __Example__
+-- For the code
+--
+-- @
+-- app todoApp {
+--   auth: {
+--     usernameAndPassword: |
+--                          ^
+--   }
+-- }
+-- @
+--
+-- The path to the cursor would be @[Decl "app", Key "auth", Key "usernameAndPassword"]@.
+type ExprPath = [ExprKey]
```

Comment:
I like the term `Path`, but I am not so sure about `Expr` -> why `Expr`? Path to Expr?  Why not CodePath? Or CursorPath? or CstPath, whatever is the best fit?